src/theory/quantifiers/quantifiers_rewriter.cpp

```diff
@@ -1743,6 +1744,90 @@ Node QuantifiersRewriter::computeVarElimination(Node body,
     }
     Trace("var-elim-quant") << "Return " << body << std::endl;
   }
+  // Leibniz equality elimination
+  if (d_opts.quantifiers.leibnizEqElim)
+  {
+    if (body.getKind() == Kind::OR && body.getNumChildren() == 2) // the body must have exactly 2 children
+    {
+      Node termA = body[0];
+      Node termB = body[1];
+
+      // helper
+      auto matchApply = [&](Node lit, Node &op, std::vector<Node> &argsOut, bool &neg) -> bool {
+        if (lit.getKind() == Kind::NOT)
+        {
+          Node in = lit[0];
+          if (in.getKind() == Kind::APPLY_UF)
+          {
+            op = in.getOperator();
+            argsOut = std::vector<Node>(in.begin(), in.end());
+            neg = true;
+            return true;
+          }
+          return false;
+        }
+        else if (lit.getKind() == Kind::APPLY_UF)
+        {
+          op = lit.getOperator();
+          argsOut = std::vector<Node>(lit.begin(), lit.end());
+          neg = false;
+          return true;
+        }
+        return false;
+      };
+
+      Node opA, opB;
+      std::vector<Node> argsA, argsB;
+      bool negA = false, negB = false;
+      if (!matchApply(termA, opA, argsA, negA) || !matchApply(termB, opB, argsB, negB))
+      {
+        return body;
+      }
+
+      // need pattern (not P(t1)) or P(t2) (either child can be the negated one)
+      if (opA != opB) return body;
```

Comment:
Can you merge these cases and ensure that `return` statements are always enclosed with `{ ... }`?